engine/access/rest/websockets/controller.go

```diff
@@ -39,62 +43,140 @@ func NewWebSocketController(
+		c.logger.Error().Err(err).Msg("error detected in one of the goroutines")
+		c.shutdownConnection()
+	}
+}
+
+// configureKeepalive sets up the WebSocket connection with a read deadline
+// and a handler for receiving pong messages from the client.
+//
+// The function does the following:
+//  1. Sets an initial read deadline to ensure the server doesn't wait indefinitely
+//     for a pong message from the client. If no message is received within the
+//     specified `pongWait` duration, the connection will be closed.
+//  2. Establishes a Pong handler that resets the read deadline every time a pong
+//     message is received from the client, allowing the server to continue waiting
+//     for further pong messages within the new deadline.
+//
+// No errors are expected during normal operation.
+func (c *Controller) configureKeepalive() error {
+	// Set the initial read deadline for the first pong message
+	// The Pong handler itself only resets the read deadline after receiving a Pong.
+	// It doesn't set an initial deadline. The initial read deadline is crucial to prevent the server from waiting
+	// forever if the client doesn't send Pongs.
+	if err := c.conn.SetReadDeadline(time.Now().Add(PongWait)); err != nil {
+		return fmt.Errorf("failed to set the initial read deadline: %w", err)
+	}
+	// Establish a Pong handler which sets the handler for pong messages received from the peer.
+	c.conn.SetPongHandler(func(string) error {
+		return c.conn.SetReadDeadline(time.Now().Add(PongWait))
+	})
+
+	return nil
 }
 
 // writeMessagesToClient reads a messages from communication channel and passes them on to a client WebSocket connection.
 // The communication channel is filled by data providers. Besides, the response limit tracker is involved in
 // write message regulation
-func (c *Controller) writeMessagesToClient(ctx context.Context) {
-	//TODO: can it run forever? maybe we should cancel the ctx in the reader routine
+//
+// Expected errors during normal operation:
+// - context.Canceled if the client disconnected
+func (c *Controller) writeMessagesToClient(ctx context.Context) error {
 	for {
 		select {
 		case <-ctx.Done():
-			return
-		case msg := <-c.communicationChannel:
+			return ctx.Err()
+		case msg, ok := <-c.communicationChannel:
+			if !ok {
+				err := fmt.Errorf("communication channel closed, no error occurred")
+				return err
+			}
 			// TODO: handle 'response per second' limits
 
+			// Specifies a timeout for the write operation. If the write
+			// isn't completed within this duration, it fails with a timeout error.
+			// SetWriteDeadline ensures the write operation does not block indefinitely
+			// if the client is slow or unresponsive. This prevents resource exhaustion
+			// and allows the server to gracefully handle timeouts for delayed writes.
+			if err := c.conn.SetWriteDeadline(time.Now().Add(WriteWait)); err != nil {
+				c.logger.Error().Err(err).Msg("failed to set the write deadline")
```

Comment:
same here and in `readMessagesFromClient`. when you have a central place were errors are handled, it's best to put the logging there instead of where each error is discovered. otherwise, it's easy to log the message multiple times.

Reply:
[5aee1ca](https://github.com/onflow/flow-go/pull/6757/commits/5aee1ca391b76d77d341b7247bbc83514d08742a)